### PR TITLE
Dev/assets download

### PR DESCRIPTION
**Pull request by Tmeister**
# Pull Request

## What changed?

New download endpoints to serve and store wp.org files.

The download endpoints serve and store core, plugins, themes, and assets files, as shown in the examples below.

```
http://api.aspiredev.org/elementor/assets/banner-1544x500.png?rev=3164133
http://api.aspiredev.org/elementor/assets/screenshot-1.gif?rev=3005087
http://api.aspiredev.org/wordpress-6.5.tar.gz
http://api.aspiredev.org/wordpress-6.2.2.tar.gz
http://api.aspiredev.org/wordpress-6.5.zip
http://api.aspiredev.org/wordpress-6.2.2.zip
http://api.aspiredev.org/theme/hello-elementor.3.1.1.zip
http://api.aspiredev.org/theme/magnum-store.1.0.3.zip
http://api.aspiredev.org/plugin/jwt-authentication-for-wp-rest-api.1.3.4.zip
```

If the file is not stored on the `SC` instance, we proxied the request to wp.org and scheduled a job to get and store a copy on `SC`; if the file is already stored, we only serve it.

Includes support for Laravel local filesystem and S3

The PR includes tests we may use to test edge cases in the future.

## Why did it change?

New featured

## Did you fix any specific issues?

Fix: #94 

## CERTIFICATION

By opening this pull request, I do agree to abide by the [CODE OF CONDUCT](https://github.com/aspirepress/.github/CODE_OF_CONDUCT.md) and be bound by the terms of the [Contribution Guidelines](https://github.com/aspirepress/.github/CONTRIBUTING.md) in effect on the date and time of my contribution as proven by the revision information in GitHub. I also agree that any previous contributions shall be deemed subject to the terms of the ersion in effect on the date and time of this pull request, or any future revisions for pull requests I may submit.

Further, I certify that this work is my own, is original, does not violate the intellectual property of any other person or entity, and I am not violating any license agreements or contracts I have with any person or entity. Finally, I agree that this code may be licensed under any license deemed appropriate by AspirePress, including but not limited to open source, closed source, proprietary or custom licenses, and that such license terms neither violate my rights or my copyright to this code.

### CODE REVIEW

File: app/Http/Controllers/API/WpOrg/Downloads/DownloadPluginController.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+     */
+    public function __invoke(string $file): Response
+    {
+        if (!\Safe\preg_match('/^([a-zA-Z0-9-]+)\.(.+)\.zip$/', $file, $matches)) {
+            return response()->json(['error' => 'Invalid plugin file format'], 400);
+        }
+
+        return $this->downloadService->download(
+            AssetType::PLUGIN_ZIP,
+            $matches[1], // plugin slug
```
Comment (chuckadams) [inline app/Http/Controllers/API/WpOrg/Downloads/DownloadPluginController.php:28]: If it's good enough to comment what it is, it's good enough to make a local variable named for what it is.

File: app/Http/Controllers/API/WpOrg/Downloads/DownloadThemeController.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+     */
+    public function __invoke(string $file): Response
+    {
+        if (!\Safe\preg_match('/^([a-zA-Z0-9-]+)\.(.+)\.zip$/', $file, $matches)) {
+            return response()->json(['error' => 'Invalid theme file format'], 400);
+        }
+
+        return $this->downloadService->download(
+            AssetType::THEME_ZIP,
+            $matches[1], // theme slug
```
Comment (chuckadams) [inline app/Http/Controllers/API/WpOrg/Downloads/DownloadThemeController.php:28]: Second verse, same as the first.  Go crazy with locals, giving things good names is half of what code is about.

File: app/Models/WpOrg/Asset.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+
+class Asset extends BaseModel
+{
+    use HasUuids;
+
+    /** @use HasFactory<AssetFactory> */
+    use HasFactory;
+
+    protected $fillable = [
```
Comment (chuckadams) [inline app/Models/WpOrg/Asset.php:18]: We also want a discriminator for the source, let's just call it `repository` and set its value to `"wp_org"`.  There should also be a created timestamp.

File: routes/api.php [hunk 4]
```diff
@@ -25,6 +29,27 @@
...
+        $router->get('/theme/{file}', DownloadThemeController::class)->where('file', '.+\.zip');
+
+        // Assets: ps.w.org/elementor/assets/screenshot-1.gif?rev=3005087
+        // Assets: ps.w.org/elementor/assets/banner-1544x500.png?rev=3164133
+        $router->get('{slug}/assets/{file}', DownloadAssetController::class)
+            ->where([
+                'slug' => '[a-zA-Z0-9-]+',
+                'file' => '.+',
+            ]);
+
```
Comment (chuckadams) [inline routes/api.php:52]: Let's put these all under a `/download` prefix

File: app/Services/Downloads/DownloadService.php
Comment (chuckadams) [inline app/Services/Downloads/DownloadService.php:1]: Could we make this just serve a redirect to the static url, whether it's upstream or a local asset?  I believe Storage::url will do the trick.  It will need to be a signed url for S3, because my trial Wasabi account doesn't support public access.  

Let's keep the streaming capability around, it could be useful for environments that want to mediate any upstream accesses, or for buckets that really need to stay private.  It should just be configurable, a setting for each of upstream and local whether to redirect or stream, with redirect by default (it's lightest on our resources and snappier for clients)

**Review by chuckadams — APPROVED** [review]

File: app/Jobs/DownloadAsset.php [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+use Illuminate\Queue\SerializesModels;
+use Illuminate\Support\Facades\Http;
+use Illuminate\Support\Facades\Storage;
+
+class DownloadAsset implements ShouldQueue
+{
+    use Dispatchable;
+    use InteractsWithQueue;
+    use Queueable;
+    use SerializesModels;
```
Comment (chuckadams) [inline app/Jobs/DownloadAsset.php:20]: Kill `SerializesModels` with fire.  Not a blocker because we're not actually using it, I just don't want jobs to have Models in their state in any way (they can use them just fine, they just have to re-query them in handle()).  I banned SerializesModels at my workplace, and lo and behold Redis stopped running out of storage...

Comment (chuckadams) [reply app/Jobs/DownloadAsset.php:20]: As for the Queueing traits, we're going to need a queue-runner container at some point so we're not running the job synchronously, but I think that can wait for a follow-up PR.

**Comment by Tmeister** [comment]
@chuckadams,

I fixed the things described in the comments.

You can see the details of the changes in the last commit message.
